File: src/Soel/Stages/AmbiguityDetector.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Soel.Stages.AmbiguityDetector
  ( detectAmbiguities
  , emitDiagnostics
  , mapIssueCategory
  ) where

import Control.Monad (unless)
import Control.Monad.Except (throwError)
import Control.Monad.IO.Class (liftIO)
import Control.Monad.Reader (ask)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.IO (stderr)

import Soel.Types (App, SoelError(..), SoelConfig(..))
import Soel.Config (CompilerConfig(..))
import Soel.IR.Types
import Soel.Utils.Logger (logInfo, formatDiagnostic, DiagnosticOpts(..))

-- | Detect ambiguities from the Narrative IR and Code IR.
-- Classifies each as error (blocks compilation) or warning (informational).
detectAmbiguities :: NarrativeIR -> CodeIR -> App AmbiguityResult
detectAmbiguities nir cir = do
  cfg <- ask
  let threshold = ccAmbiguityThreshold (cfgCompiler cfg)

      (a1, idx1) = detectLowConfidenceEntities threshold (nirEntities nir) 1
      (a2, idx2) = detectConflictingRelations (nirRelationships nir) idx1
      (a3, idx3) = detectNarrativeAmbiguities threshold (nirAmbiguities nir) idx2
      (a4, idx4) = detectMissingEntryPoint (cirEntryPoint cir) idx3
      (a5, _)    = detectLowConfidenceEvents threshold (nirEvents nir) idx4

      allAmbs  = a1 ++ a2 ++ a3 ++ a4 ++ a5
      errors   = filter (\a -> ambSeverity a == SevError) allAmbs
      warnings = filter (\a -> ambSeverity a == SevWarning) allAmbs

  unless (null errors) $
    liftIO $ logInfo $ T.pack (show (length errors)) <> " semantic error(s), "
                    <> T.pack (show (length warnings)) <> " warning(s)"

  unless (not (null errors) || null warnings) $
    liftIO $ logInfo $ T.pack (show (length warnings)) <> " semantic warning(s)"

  pure AmbiguityResult
    { arErrors   = errors
    , arWarnings = warnings
    , arAll      = allAmbs
    }

-- ─── Pure detection functions ─────────────────────────────────────

mkId :: Int -> Text
mkId i = "S" <> T.justifyRight 3 '0' (T.pack (show i))

-- | Detect low-confidence entity mentions.
detectLowConfidenceEntities :: Double -> [NarrativeEntity] -> Int -> ([Ambiguity], Int)
detectLowConfidenceEntities threshold entities startIdx =
  foldr go ([], startIdx) entities
  where
    go entity (acc, idx) =
      let lowConf = filter (\m -> menConfidence m < threshold) (neMentions entity)
      in if null lowConf
         then (acc, idx)
         else
           let maxConf   = maximum (map menConfidence lowConf)
               sev       = if maxConf < threshold * 0.5 then SevError else SevWarning
               firstSpan = menSpan (head lowConf)
               amb = Ambiguity
                 { ambId          = mkId idx
                 , ambSeverity    = sev
                 , ambCategory    = CatNaming
                 , ambDescription = "Ambiguous entity \"" <> neCanonicalName entity
                                 <> "\" -- semantic encoder cannot confidently resolve this concept"
                 , ambSourceSpan  = Just firstSpan
                 , ambOptions     =
                     [ AmbiguityOption
                         { aoLabel       = "Keep as \"" <> neCanonicalName entity <> "\""
                         , aoDescription = "Use the name as-is for the Haskell type"
                         , aoConfidence  = 0.6
                         }
                     , AmbiguityOption
                         { aoLabel       = "Rename / clarify"
                         , aoDescription = "Provide a clearer name for this concept"
                         , aoConfidence  = 0.4
                         }
                     ]
                 , ambResolution  = Nothing
                 }
           in (amb : acc, idx + 1)

-- | Detect conflicting relationships (multiple rels between same entity pair).
detectConflictingRelations :: [NarrativeRelationship] -> Int -> ([Ambiguity], Int)
detectConflictingRelations rels startIdx =
  let relMap :: Map (Text, Text) [NarrativeRelationship]
      relMap = Map.fromListWith (++)
        [ ((nrSourceEntityId r, nrTargetEntityId r), [r]) | r <- rels ]
      conflicts = filter (\(_, rs) -> length rs > 1) (Map.toList relMap)
  in foldr go ([], startIdx) conflicts
  where
    go (key, rs) (acc, idx) =
      let names  = map (\r -> "\"" <> nrRelation r <> "\"") rs
          keyStr = fst key <> "-" <> snd key
          amb = Ambiguity
            { ambId          = mkId idx
            , ambSeverity    = SevError
            , ambCategory    = CatRelation
            , ambDescription = "Conflicting relations between " <> keyStr <> ": "
                            <> T.intercalate " vs " names
            , ambSourceSpan  = Just (nrEvidenceSpan (head rs))
            , ambOptions     = map (\r -> AmbiguityOption
                { aoLabel       = nrRelation r
                , aoDescription = "Use \"" <> nrRelation r <> "\" relationship"
                , aoConfidence  = nrConfidence r
                }) rs
            , ambResolution  = Nothing
            }
      in (amb : acc, idx + 1)

-- | Detect narrative-level ambiguities from the encoder.
detectNarrativeAmbiguities :: Double -> [NarrativeAmbiguity] -> Int -> ([Ambiguity], Int)
detectNarrativeAmbiguities threshold narAmbs startIdx =
  foldr go ([], startIdx) narAmbs
  where
    go na (acc, idx) =
      let interps = naInterpretations na
          maxConf = if null interps then 0.0
                    else maximum (map intConfidence interps)
          spread  = if length interps > 1
                    then abs (intConfidence (head interps) - intConfidence (interps !! 1))
                    else 1.0
          sev = if spread < 0.2 || maxConf < threshold then SevError else SevWarning
          amb = Ambiguity
            { ambId          = mkId idx
            , ambSeverity    = sev
            , ambCategory    = mapIssueCategory (naIssue na)
            , ambDescription = naIssue na
            , ambSourceSpan  = Just (naSpan na)
            , ambOptions     = map (\interp -> AmbiguityOption
                { aoLabel       = T.take 80 (intReading interp)
                , aoDescription = intReading interp
                , aoConfidence  = intConfidence interp
                }) interps
            , ambResolution  = Nothing
            }
      in (amb : acc, idx + 1)

-- | Detect missing entry point.
detectMissingEntryPoint :: Maybe EntryPointDecl -> Int -> ([Ambiguity], Int)
detectMissingEntryPoint (Just _) idx = ([], idx)
detectMissingEntryPoint Nothing  idx =
  ( [ Ambiguity
      { ambId          = mkId idx
      , ambSeverity    = SevError
      , ambCategory    = CatBehavior
      , ambDescription = "No entry point: program has no discernible main behavior"
      , ambSourceSpan  = Nothing
      , ambOptions     =
          [ AmbiguityOption
              { aoLabel       = "Interactive CLI"
              , aoDescription = "Run as an interactive command-line program"
              , aoConfidence  = 0.3
              }
          , AmbiguityOption
              { aoLabel       = "Print demo"
              , aoDescription = "Print a demonstration of the defined types and functions"
              , aoConfidence  = 0.5
              }
          , AmbiguityOption
              { aoLabel       = "Custom"
              , aoDescription = "Let me describe the entry point"
              , aoConfidence  = 0.2
              }
          ]
      , ambResolution  = Nothing
      }
    ]
  , idx + 1
  )

-- | Detect low-confidence events.
detectLowConfidenceEvents :: Double -> [NarrativeEvent] -> Int -> ([Ambiguity], Int)
detectLowConfidenceEvents threshold events startIdx =
  foldr go ([], startIdx) events
  where
    go event (acc, idx) =
      let cert = modCertainty (nevModality event)
      in if cert >= threshold
         then (acc, idx)
         else
           let sev = if cert < threshold * 0.5 then SevError else SevWarning
               amb = Ambiguity
                 { ambId          = mkId idx
                 , ambSeverity    = sev
                 , ambCategory    = CatBehavior
                 , ambDescription = "Uncertain semantics for \"" <> nevPredicate event
                                 <> "\" -- cannot determine intended behavior"
                 , ambSourceSpan  = Just (etSpan (nevTrigger event))
                 , ambOptions     =
                     [ AmbiguityOption
                         { aoLabel       = "Keep as described"
                         , aoDescription = "Implement \"" <> nevPredicate event
                                        <> "\" as the encoder understood it"
                         , aoConfidence  = cert
                         }
                     , AmbiguityOption
                         { aoLabel       = "Clarify behavior"
                         , aoDescription = "Provide more details about what this should do"
                         , aoConfidence  = 1 - cert
                         }
                     ]
                 , ambResolution  = Nothing
                 }
           in (amb : acc, idx + 1)

-- | Emit diagnostics to stderr and throw if there are unresolved errors.
emitDiagnostics :: AmbiguityResult -> FilePath -> Text -> App ()
emitDiagnostics result filePath sourceText = do
  liftIO $ mapM_ (\amb -> TIO.hPutStr stderr $ formatDiagnostic $ ambiguityDiag filePath sourceText amb)
    (arAll result)

  unless (null (arErrors result)) $
    throwError $ SemanticAmbiguityError $
      "Compilation failed: " <> T.pack (show (length (arErrors result)))
      <> " semantic error(s), " <> T.pack (show (length (arWarnings result)))
      <> " warning(s)"

-- | Build diagnostic options from an ambiguity.
ambiguityDiag :: FilePath -> Text -> Ambiguity -> DiagnosticOpts
ambiguityDiag fp sourceText amb = DiagnosticOpts
  { diagFile       = fp
  , diagSeverity   = severityText (ambSeverity amb)
  , diagId         = ambId amb
  , diagCategory   = categoryText (ambCategory amb)
  , diagMessage    = ambDescription amb
  , diagSpan       = ambSourceSpan amb
  , diagSourceText = Just sourceText
  , diagOptions    = map (\o -> (aoLabel o, aoConfidence o)) (ambOptions amb)
  }

-- | Map issue text to an ambiguity category.
mapIssueCategory :: Text -> AmbiguityCategory
mapIssueCategory issue
  | T.isInfixOf "type" issue || T.isInfixOf "kind" issue = CatType
  | T.isInfixOf "scope" issue                             = CatScope
  | T.isInfixOf "name" issue || T.isInfixOf "coref" issue = CatNaming
  | T.isInfixOf "relation" issue || T.isInfixOf "causal" issue = CatRelation
  | T.isInfixOf "constrain" issue || T.isInfixOf "rule" issue  = CatConstraint
  | otherwise                                              = CatOther

File: src/stages/ambiguity-detector.ts
import type { NarrativeIR, CodeIR } from '../ir/types.js';
import type { Ambiguity, AmbiguitySeverity } from '../ir/types.js';
import type { SoelConfig } from '../config.js';
import { SemanticAmbiguityError } from '../utils/errors.js';
import { log, formatDiagnostic } from '../utils/logger.js';

export interface AmbiguityResult {
  errors: Ambiguity[];
  warnings: Ambiguity[];
  all: Ambiguity[];
}

/**
 * Detect ambiguities from the Narrative IR and Code IR.
 * Classifies each as error (blocks compilation) or warning (informational).
 */
export function detectAmbiguities(
  narrativeIR: NarrativeIR,
  codeIR: CodeIR,
  config: SoelConfig
): AmbiguityResult {
  const threshold = config.compiler.ambiguityThreshold;
  const ambiguities: Ambiguity[] = [];
  let idx = 1;

  // 1. Low-confidence entities → naming/type ambiguity
  for (const entity of narrativeIR.entities) {
    const lowConf = entity.mentions.filter((m) => m.confidence < threshold);
    if (lowConf.length > 0) {
      const maxConf = Math.max(...lowConf.map((m) => m.confidence));
      ambiguities.push({
        id: `S${String(idx++).padStart(3, '0')}`,
        severity: maxConf < threshold * 0.5 ? 'error' : 'warning',
        category: 'naming',
        description: `Ambiguous entity "${entity.canonical_name}" — semantic encoder cannot confidently resolve this concept`,
        sourceSpan: lowConf[0].span,
        options: [
          {
            label: `Keep as "${entity.canonical_name}"`,
            description: 'Use the name as-is for the Haskell type',
            confidence: 0.6,
          },
          {
            label: 'Rename / clarify',
            description: 'Provide a clearer name for this concept',
            confidence: 0.4,
          },
        ],
      });
    }
  }

  // 2. Conflicting relationships → always error
  const relMap = new Map<string, typeof narrativeIR.relationships>();
  for (const rel of narrativeIR.relationships) {
    const key = `${rel.source_entity_id}-${rel.target_entity_id}`;
    const existing = relMap.get(key) ?? [];
    existing.push(rel);
    relMap.set(key, existing);
  }
  for (const [key, rels] of relMap) {
    if (rels.length > 1) {
      const names = rels.map((r) => `"${r.relation}"`);
      ambiguities.push({
        id: `S${String(idx++).padStart(3, '0')}`,
        severity: 'error',
        category: 'relation',
        description: `Conflicting relations between ${key}: ${names.join(' vs ')}`,
        sourceSpan: rels[0].evidence_span,
        options: rels.map((r) => ({
          label: r.relation,
          description: `Use "${r.relation}" relationship`,
          confidence: r.confidence,
        })),
      });
    }
  }

  // 3. Narrative-level ambiguities from the encoder
  for (const amb of narrativeIR.ambiguities) {
    const maxConf = Math.max(...amb.interpretations.map((i) => i.confidence), 0);
    const spread = amb.interpretations.length > 1
      ? Math.abs(amb.interpretations[0].confidence - amb.interpretations[1].confidence)
      : 1;
    // Error if top interpretations are too close in confidence (genuine ambiguity)
    // or if even the best reading is low confidence
    const sev: AmbiguitySeverity = (spread < 0.2 || maxConf < threshold) ? 'error' : 'warning';

    ambiguities.push({
      id: `S${String(idx++).padStart(3, '0')}`,
      severity: sev,
      category: mapIssueCategory(amb.issue),
      description: amb.issue,
      sourceSpan: amb.span,
      options: amb.interpretations.map((interp) => ({
        label: interp.reading.slice(0, 80),
        description: interp.reading,
        confidence: interp.confidence,
      })),
    });
  }

  // 4. Missing entry point → error
  if (!codeIR.entryPoint) {
    ambiguities.push({
      id: `S${String(idx++).padStart(3, '0')}`,
      severity: 'error',
      category: 'behavior',
      description: 'No entry point: program has no discernible main behavior',
      options: [
        {
          label: 'Interactive CLI',
          description: 'Run as an interactive command-line program',
          confidence: 0.3,
        },
        {
          label: 'Print demo',
          description: 'Print a demonstration of the defined types and functions',
          confidence: 0.5,
        },
        {
          label: 'Custom',
          description: 'Let me describe the entry point',
          confidence: 0.2,
        },
      ],
    });
  }

  // 5. Low-confidence events → severity depends on how uncertain
  for (const event of narrativeIR.events) {
    if (event.modality.certainty < threshold) {
      ambiguities.push({
        id: `S${String(idx++).padStart(3, '0')}`,
        severity: event.modality.certainty < threshold * 0.5 ? 'error' : 'warning',
        category: 'behavior',
        description: `Uncertain semantics for "${event.predicate}" — cannot determine intended behavior`,
        sourceSpan: event.trigger.span,
        options: [
          {
            label: 'Keep as described',
            description: `Implement "${event.predicate}" as the encoder understood it`,
            confidence: event.modality.certainty,
          },
          {
            label: 'Clarify behavior',
            description: 'Provide more details about what this should do',
            confidence: 1 - event.modality.certainty,
          },
        ],
      });
    }
  }

  const errors = ambiguities.filter((a) => a.severity === 'error');
  const warnings = ambiguities.filter((a) => a.severity === 'warning');

  if (errors.length > 0) {
    log.info(`${errors.length} semantic error(s), ${warnings.length} warning(s)`);
  } else if (warnings.length > 0) {
    log.info(`${warnings.length} semantic warning(s)`);
  }

  return { errors, warnings, all: ambiguities };
}

/**
 * Emit diagnostics to stderr and throw if there are unresolved errors.
 * This is the "strict" path — ambiguities are compiler errors.
 */
export function emitDiagnostics(
  result: AmbiguityResult,
  filePath: string,
  sourceText: string
): void {
  for (const amb of result.all) {
    console.error(formatDiagnostic({
      file: filePath,
      severity: amb.severity,
      id: amb.id,
      category: amb.category,
      message: amb.description,
      span: amb.sourceSpan,
      sourceText,
      options: amb.options,
    }));
  }

  if (result.errors.length > 0) {
    const summary = `Compilation failed: ${result.errors.length} semantic error(s), ${result.warnings.length} warning(s)`;
    throw new SemanticAmbiguityError(summary, '');
  }
}

function mapIssueCategory(issue: string): Ambiguity['category'] {
  if (issue.includes('type') || issue.includes('kind')) return 'type';
  if (issue.includes('scope')) return 'scope';
  if (issue.includes('name') || issue.includes('coref')) return 'naming';
  if (issue.includes('relation') || issue.includes('causal')) return 'relation';
  if (issue.includes('constrain') || issue.includes('rule')) return 'constraint';
  return 'other';
}
